kubernetes/samples/scripts/domain-lifecycle/helper.sh

```diff
@@ -0,0 +1,606 @@
+    prefix=$(echo ${dynaCluster} | jq -r .serverNamePrefix)
+    clusterSize=$(echo ${dynaCluster} | jq .dynamicClusterSize) 
+    for (( i=1; i<=$clusterSize; i++ )); do
+      localServerName=${prefix}$i
+      sortedServers+=(${localServerName})
+    done
+  fi
+  # Create arrays of ALWAYS policy servers and other servers
+  sortedServersSize=${#sortedServers[@]}
+  if [ "${sortedServersSize}" -gt 0 ]; then
+    for localServerName in "${sortedServers[@]}"; do
+      getEffectivePolicy "${domainJson}" "${localServerName}" "${clusterName}" policy
+      # Update policy when server name matches current server and unsetting
+      if [[ "${withPolicy}" == "UNSET" && "${serverName}" == "${localServerName}" ]]; then
+        policy=UNSET
+      fi
+      if [ "${policy}" == "ALWAYS" ]; then
+        sortedByAlwaysServers+=(${localServerName})
+      else
+        otherServers+=(${localServerName})
+      fi
+    done
+  fi
+  
+  # append other servers to the list of servers with always policy
+  otherServersSize=${#otherServers[@]}
+  if [ "${otherServersSize}" -gt 0 ]; then
+    for otherServer in "${otherServers[@]}"; do
+      sortedByAlwaysServers+=($otherServer)
+    done
+  fi
+}
+
+#
+# Get replica count for a cluster
+# $1 - Domain resource in json format
+# $2 - Name of cluster 
+# $3 - Return value containing replica count
+#
+function getReplicaCount {
+  local domainJson=$1
+  local clusterName=$2
+  local __replicaCount=$3
+
+  replicasCmd="(.spec.clusters[] \
+    | select (.clusterName == \"${clusterName}\")).replicas"
+  replicaCount=$(echo ${domainJson} | jq "${replicasCmd}")
+  eval $__replicaCount="'${replicaCount}'"
+
+}
+
+#
+# Check servers started in a cluster based on server start policy and 
+# replica count.
+# $1 - Domain resource in json format
+# $2 - Name of server 
+# $3 - Name of cluster 
+# $4 - Indicates if replicas will stay constant, incremented or decremented
+# $5 - Indicates if policy of current server will stay same or unset
+# $6 - Return value indicating if current server will be started
+#
+function checkStartedServers {
+  local domainJson=$1
+  local serverName=$2
+  local clusterName=$3
+  local withReplicas=$4
+  local withPolicy=$5
+  local __started=$6
+  local localServerName=""
+  local policy=""
+  local replicaCount=0
+  local currentReplicas=0
+  local startedServers=()
+  local sortedByAlwaysServers=()
+  
+  # Get sorted list of servers in 'sortedByAlwaysServers' array
+  getSortedListOfServers "${domainJson}" "${serverName}" "${clusterName}" "${withPolicy}"
+  getReplicaCount "${domainJson}" "${clusterName}" replicaCount
+  # Increment or decrement the replica count based on 'withReplicas' input parameter
+  if [ "${withReplicas}" == "INCREASED" ]; then
+    replicaCount=$((replicaCount+1))
+  elif [ "${withReplicas}" == "DECREASED" ]; then
+    replicaCount=$((replicaCount-1))
+  fi
+  sortedByAlwaysSize=${#sortedByAlwaysServers[@]}
+  if [ "${sortedByAlwaysSize}" -gt 0 ]; then
```

Comment:
just a minor point:  in my experimentation changing the following:
```
    # note that the lack of quotes is deliberate
    for localServerName in ${sortedByAlwaysServers[@]:-}; do
    done
```
is equivalent to:
```
  sortedByAlwaysSize=${#sortedByAlwaysServers[@]}
  if [ "${sortedByAlwaysSize}" -gt 0 ]; then
    for localServerName in "${sortedByAlwaysServers[@]}"; do
    done
  fi
```
The drawback is that it  (a) assumes server names won't have special chars like quotes in them - which is probably a reasonable assumption, and (b) it defeats the 'set -u' var name check so a bad var name would short circuit the loop...

Reply:
Changed to use new syntax to avoid size check for empty arrays.